### PR TITLE
Smart computer

### PR DESCRIPTION
**Pull request by Gabbendorf**
Adds unbeatable computer to the game. Unbeatable computer can only be codebreaker (not codemaker), and it guesses correctly within 8 turns (usually within 4/5 guesses).
At the end of the game, user is asked whether they want to play again.

Unbeatable computer implements methods that reflect following algorithm:
- first guess is random and made up of 2 different colours repeated twice
- previous guess becomes computer's internal code pattern (regardless the feedback given by board)
- next guess is chosen among a list of all possible patterns. After previous guess, computer compares each of these possible patterns with its internal code pattern against feedback returned. If feedback for each of those is different from feedback returned by the board for computer's guess, those possible patterns are deleted from the list. This operation drastically reduces list of possible patterns (so of possible future guesses) at each turn.
- Last guess becomes new internal code pattern against which remained possible patterns are compared for next turn.
- Above cycle (apart from step 1) is repeated until computer guesses secret code pattern.

### CODE REVIEW

File: spec/unbeatable_computer_spec.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    second_colour = second_guess.colours[1].colour
+    third_colour = second_guess.colours[2].colour
+    fourth_colour = second_guess.colours[3].colour
+
+    second_guess_colours = second_guess.colours.map(&:colour)
+    expect(second_guess_colours).to eq([first_colour, second_colour, third_colour, fourth_colour])
+  end
+
+  it "guesses correctly within given number of guesses" do
+    def make_guess_until_win
```
Comment (RabeaGleissner) [inline spec/unbeatable_computer_spec.rb]: You normally wouldn't define a method inside a test. Maybe inside a context but normally just outside both test and context.

File: lib/board.rb [hunk 2]
```diff
@@ -20,6 +19,12 @@ def printable_history(result)
     }
   end
 
+  def show_guesses_and_feedback
+    feedback = {}
+    @history.each {|result| feedback[result.guess] = result.feedback}
+    feedback
```
Comment (RabeaGleissner) [inline lib/board.rb]: inconsistent with method name - it returns just "feedback"

File: lib/players.rb [hunk 3]
```diff
@@ -9,20 +10,22 @@ def initialize(ui)
...
 
-  def codebreaker(input)
-    if input == "computer"
+  def codebreaker(input, board)
+    if input == "c"
       Computer.new("computer-codebreaker", 4)
-    elsif input == "human player"
+    elsif input == "h"
       HumanPlayer.new(@ui.ask_human_player_name, @ui, 4)
+    elsif input == "s"
```
Comment (RabeaGleissner) [inline lib/players.rb:27]: inconsistent naming here - "smart computer" vs. "unbeatable computer". Pick one and stick to it

**Review by RabeaGleissner — COMMENTED** [review]
let's chat about it in the IPM!

File: lib/mastermind.rb [hunk 4]
```diff
@@ -43,6 +43,8 @@ def next_move(board, pattern, codemaker, codebreaker)
   def end_of_game(board, codemaker, codebreaker)
     game_verdict = board.verdict
     game_verdict == :codemaker_wins ? @ui.codemaker_is_winner(codemaker.name) : @ui.codebreaker_is_winner(codebreaker.name)
+    answer = @ui.play_again
+    answer == "y" ? run_game : @ui.say_goodbye
```
Comment (Gabbendorf) [inline lib/mastermind.rb]: No need for the temporary variable here

File: spec/board_spec.rb [hunk 5]
```diff
@@ -53,6 +53,17 @@ def set_up_result(red_pegs, white_pegs)
...
 
+  it "returns hash with all feedback from history" do
+    first_result = set_up_result(1, 2)
+    second_result = set_up_result(1, 2)
+    board.keep_track_of_results(first_result)
+    board.keep_track_of_results(second_result)
+
+    hash_of_feedback = board.show_guesses_and_feedback
+
+    expect(hash_of_feedback).to eq({first_result.guess => first_result.feedback, second_result.guess => second_result.feedback})
```
Comment (Gabbendorf) [inline spec/board_spec.rb]: Inline board.show_guesses_and_feedback and save expected result in variable

File: spec/players_spec.rb [hunk 6]
```diff
@@ -15,15 +20,15 @@
   end
 
   it "creates computer codemaker" do
-    input = "computer"
+    input = "c"
```
Comment (Gabbendorf) [inline spec/players_spec.rb]: This variable should be more descriptive (for ex player_selection)

File: spec/ui_spec.rb [hunk 7]
```diff
@@ -26,51 +27,51 @@
...
+    codebreaker = ui.choose_codebreaker(players, board)
 
-    expect(output.string).to eq("Select the codebreaker (computer or human player)\nEnter player's name:\n")
+    expect(output.string).to eq("Select the codebreaker ('c' = computer, 'h' = human player or 's' = smart computer)\nEnter player's name:\n")
     expect(codebreaker.name).to eq("Gabi")
   end
 
   it "asks for name for human player" do
-    input = StringIO.new("Gabriella\n")
+    input = StringIO.new("Gabriella")
```
Comment (Gabbendorf) [inline spec/ui_spec.rb]: NO need to create a variable of input

File: lib/unbeatable_computer.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  end
+
+  def make_next_guess
+    @all_possible_patterns = compatible_patterns
+    guess = @all_possible_patterns.sample
+    update_possible_patterns_and_temporary_pattern(guess)
+    guess
+  end
+
+  def generate_all_possible_patterns
```
Comment (Gabbendorf) [inline lib/unbeatable_computer.rb]: You can use ruby method

**Pull request merged**